tests/run/gil_in_var_initialization_tests.pyx

```diff
@@ -0,0 +1,137 @@
+# mode: run
+# tag: cpp, cpp11
+
+from libcpp cimport bool
+
+# What we're testing in this file is that certain bits of code don't acquire the GIL
+# at any point during the function.
+#
+# To do this we create a C++ thread that runs with the GIL permanently held, and
+# which is waiting for a condition variable to be set. We run the relevant Cython
+# function and then set the condition variable.
+#
+# If the Cython function requires the GIL it will be blocked by the C++ thread and
+# so be unable to set the condition variable.
+#
+# This isn't specifically a C++ test, but the C++ standard library concurrency tools
+# are a convenient way of getting condition variables and mutexes.
+
+cdef extern from *:
+    """
+    #include <condition_variable>
+    #include <mutex>
+    #include <chrono>
+    #include <future>
+
+    static bool waiting = false;
+    static bool triggered = false;
+    static std::condition_variable condition_variable;
+    static std::mutex mutex;
+
+    void reset_cpp_state() {
+        std::unique_lock<std::mutex> lock(mutex);
+        waiting = false;
+        triggered = false;
+    }
+
+    bool block_and_wait_for_trigger() {
+        std::unique_lock<std::mutex> lock(mutex);
+        waiting = true;
+        triggered = false;
+        condition_variable.notify_all();
+        // Note - I'm using a 5 second timeout here to avoid the test
+        // being "deadlock forever". This could occasionally give false
+        // failures if things are running really slowly
```

Comment:
Yeah, that would be annoying. I dislike uncontrolled concurrency in tests for pretty much this reason. Threads are one of those tools that programmers easily consider appropriate and that rarely are. Especially if test failures necessarily make the code run into timeouts, rather than failing timely.

Some people, when confronted with a problem, say "I know, I'll use threads". And now problems two have they. :)

Can't we use a C code validation check that errors if it finds any refnanny usage before the module init function, or something like that? That seems doable with a regex. (Which, admittedly, is the other tool which gives you two problems from one for free…)

We could also extend the C code validation to allow start and end markers, to match a specific function or section. `sed` uses `/startregex/,/endregex/ command`, for example, where start and end are optional. For our use case, we could use something like `/startregex/,/endregex/   innerregex`, and fail if the inner regex is (not) found, as well as if start or end are not found (because we always expect those to be present either way).

Reply:
Improved C code validation is implemented in https://github.com/cython/cython/commit/09bb9e1208a9dff2305d6bbba8803823b6386e05

Pattern format is `/start/:/end/pattern`, or something like `/start/ : /end/  pattern` if you prefer adding whitespace. Both start and end are optional.

Reply:
I'll add C code validation.

I'd like to leave this test in though - I think it is the "right" test and making sure the actual problem doesn't recur.. I think the only thing that is non-deterministic is the slight chance it takes more than 5 seconds. But realistically that's a huge statement of the amount of time it needs

Reply:
I've added some C code tests. I think we should skip the more complicated patterns on annotated HTML though - they seem almost impossible to make work there